packages/pyright-internal/src/analyzer/service.ts

```diff
@@ -483,9 +500,16 @@ export class AnalyzerService {
     }
 
     protected runAnalysis(token: CancellationToken) {
-        const moreToAnalyze = this._backgroundAnalysisProgram.startAnalysis(token);
-        if (moreToAnalyze) {
-            this._scheduleReanalysis(/* requireTrackedFileUpdate */ false);
+        // In pull diagnostics mode, ask VS code to refresh the diagnostics. This is how
+        // we tell VS code that something happened that might affect diagnostics.
+        if (this.options.usingPullDiagnostics) {
+            this._connection?.sendRequest(DiagnosticRefreshRequest.type);
```

Comment:
won't this cause all open files to get pull diagnostics request for every char changes?

Reply:
I believe VS code is smart enough to only queue up 1 request after a change happens, but let me double check

Reply:
No, you're right. It causes double requests. We don't actually compute anything on the second request (as the previous result id matches) but it's still sending them.

Reply:
Oh actually this is already called during the timer

Reply:
I changed it to be a callback and only happen during 'invalidateAndForceReanalysis'. This should handle the cases that VS code doesn't already cause an update for